Mathlib/Analysis/SpecialFunctions/Log/Base.lean

```diff
@@ -427,3 +436,67 @@ theorem logb_prod {α : Type*} (s : Finset α) (f : α → ℝ) (hf : ∀ x ∈
 #align real.logb_prod Real.logb_prod
 
 end Real
+
+section Induction
+
+open Real in
+/-- Induction principle for intervals of real numbers: if a proposition `P` is true
+on `[x₀, r * x₀)` and if `P` for `[x₀, r^n * x₀)` implies `P` for `[r^n * x₀, r^(n+1) * x₀)`,
+then `P` is true for all `x ≥ x₀`. -/
+lemma Real.induction_Ico_mul {P : ℝ → Prop} (x₀ r : ℝ) (hr : 1 < r) (hx₀ : 0 < x₀)
+    (base : ∀ x ∈ Set.Ico x₀ (r * x₀), P x)
+    (step : ∀ n : ℕ, n ≥ 1 → (∀ z ∈ Set.Ico x₀ (r ^ n * x₀), P z) →
+      (∀ z ∈ Set.Ico (r ^ n * x₀) (r ^ (n+1) * x₀), P z)) :
+    ∀ x ≥ x₀, P x :=
+  fun x hx =>
+    have h₁ : r ≠ 1 := (ne_of_lt hr).symm
+    have h₂ : 0 < r := zero_lt_one.trans hr
+    have h₃ : x₀ ≠ 0 := (ne_of_lt hx₀).symm
+    have h₄ : 0 < x := by linarith
+    if hx' : x < r * x₀ then
+      base x ⟨hx, hx'⟩
+    else by
+      push_neg at hx'
+      refine step ⌊logb r (x / x₀)⌋₊ ?ge_one ?main x ?memIco
+      case ge_one =>
+        calc 1 = ⌊logb r (r * x₀ / x₀)⌋₊ := by
+                simp [mul_div_cancel _ (ne_of_lt hx₀).symm, hr]
+             _ ≤ ⌊logb r (x / x₀)⌋₊ := by gcongr; exact hr
+      case main =>
+        intro z ⟨z_lb, z_ub⟩
+        have z_pos : 0 < z := by linarith
+        have h_nonneg : 0 ≤ logb r (z / x₀) := logb_nonneg hr ((one_le_div hx₀).mpr z_lb)
+        have hlogb : logb r (r ^ ⌊logb r (x / x₀)⌋₊ * x₀ / x₀) =  ⌊logb r (x / x₀)⌋₊ := by
+          rw [mul_div_cancel _ (ne_of_lt hx₀).symm, logb_rpow h₂ h₁]
+        -- Needed by the termination checker
+        have _well_founded : ⌊logb r (z / x₀)⌋₊ < ⌊logb r (x / x₀)⌋₊ := by
+          rw [Nat.floor_lt h_nonneg]
+          calc logb r (z / x₀) < logb r (r ^ ⌊logb r (x / x₀)⌋₊ * x₀ / x₀) := by
+                    refine (logb_lt_logb_iff hr (div_pos z_pos hx₀) (by positivity)).mpr ?_
+                    gcongr
+               _ = ⌊logb r (r ^ ⌊logb r (x / x₀)⌋₊ * x₀ / x₀)⌋₊ := by
+                    nth_rewrite 1 [hlogb]
+                    norm_cast
+                    rw [←rpow_nat_cast, hlogb]
+                    simp
+               _ = ⌊logb r (x / x₀)⌋₊ := by simp only [hlogb, Nat.floor_coe]
+        exact induction_Ico_mul x₀ r hr hx₀ base step z z_lb
+      case memIco =>
+        have h₅ : x = r ^ logb r (x / x₀) * x₀ := by
+          rw [rpow_logb h₂ h₁ (by positivity), div_mul_cancel _ h₃]
+        refine ⟨?lb, ?ub⟩
+        case lb =>
+          calc _ ≤ r ^ logb r (x / x₀) * x₀ := by
+                  gcongr
+                  · exact le_of_lt hr
+                  · exact Nat.floor_le (logb_nonneg hr (by rwa [one_le_div hx₀]))
+               _ = x := by rw [←h₅]
+        case ub =>
+          calc _ = r ^ logb r (x / x₀) * x₀ := by rw [←h₅]
+               _ < r ^ (⌊logb r (x / x₀)⌋₊ + 1) * x₀ := by
+                  gcongr
+                  simp [rpow_lt_rpow_left_iff hr, Nat.lt_floor_add_one]
+  termination_by induction_Ico_mul x₀ r hr hx₀ base step x hx => ⌊logb r (x / x₀)⌋₊
```

Comment:
```suggestion
/-- Induction principle for intervals of real numbers: if a proposition `P` is true
on `[x₀, r * x₀)` and if `P` for `[x₀, r^n * x₀)` implies `P` for `[r^n * x₀, r^(n+1) * x₀)`,
then `P` is true for all `x ≥ x₀`. -/
lemma Real.induction_Ico_mul {P : ℝ → Prop} (x₀ r : ℝ) (hr : 1 < r) (hx₀ : 0 < x₀)
    (base : ∀ x ∈ Set.Ico x₀ (r * x₀), P x)
    (step : ∀ n : ℕ, n ≥ 1 → (∀ z ∈ Set.Ico x₀ (r ^ n * x₀), P z) →
      (∀ z ∈ Set.Ico (r ^ n * x₀) (r ^ (n+1) * x₀), P z)) :
    ∀ x ≥ x₀, P x := by
  suffices : ∀ n : ℕ, ∀ x ∈ Set.Ico x₀ (r ^ (n + 1) * x₀), P x
  · intro x hx
    have hx' : 0 < x / x₀ := div_pos (hx₀.trans_le hx) hx₀
    refine this ⌊logb r (x / x₀)⌋₊ x ?_
    rw [mem_Ico, ←div_lt_iff hx₀, ←logb_lt_iff_lt_rpow hr hx']
    exact ⟨hx, Nat.lt_floor_add_one _⟩
  intro n
  induction n
  case zero => simpa using base
  case succ n ih =>
    specialize step (n + 1) (by simp)
    simp only [Nat.cast_add_one] at step ⊢
    exact fun x hx => (Ico_subset_Ico_union_Ico hx).elim (ih x) (step ih _)
```
You can simplify this proof quite a bit!

Reply:
If you'd like you can also make it even shorter and more in term mode. But this one is optional:
```
  have ind : ∀ n : ℕ, ∀ x ∈ Set.Ico x₀ (r ^ (n + 1) * x₀), P x :=
    Nat.rec (fun x hx => base x <| by simpa using hx) fun n ih x hx =>
      (Ico_subset_Ico_union_Ico hx).elim (ih x)
        fun hx' => step (n + 1) (by simp) (by simpa using ih) _ (by simpa using hx')
  fun x hx =>
    ind ⌊logb r (x / x₀)⌋₊ x <| mem_Ico.2 <|
      have hx' : 0 < x / x₀ := div_pos (hx₀.trans_le hx) hx₀
      ⟨hx, (div_lt_iff hx₀).1 <| (logb_lt_iff_lt_rpow hr hx').1 <| Nat.lt_floor_add_one _⟩
```

Reply:
Very nice, thanks!!